create additional index on task table

pranavbhole opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

Fixes #XXXX.

<!-- Replace XXXX with the id of the issue fixed in this PR. Remove this section if there is no corresponding issue. Don't reference the issue in the title of this pull-request. -->

<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description
We identified the issue with the getTasks api for given datasource while investigating the high CPU on metadata.  getTasks api runs the query SELECT `id` , `status_payload` , `payload` , `datasource` , `created_date` FROM `druid_tasks` WHERE `active` = TRUE AND `datasource` = ? ORDER BY `created_date`  and scan 50% task table due current index.  We should create additional index on the task table to speed up this api. 
This PR will create the additional index on task table with retry if not already created. Added the test cases and log to make sure that index creation never fails.


<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->


<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->


<hr>

##### Key changed/added classes in this PR
 * `SQLMetadataConnector`

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

suneet-s commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -377,8 +380,20 @@ public void createEntryTable(final String tableName)
+    );
+    createIndex(
+        tableName,
+        "active_created",
```
```suggestion
        "active_created_date",
```

pranavbhole replied:
fixed

suneet-s commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -830,4 +845,73 @@ public Void withHandle(Handle handle)
+            public Void withHandle(Handle handle)
+            {
+              if (tableExists(handle, tableName)) {
+                Map<String, List<String>> indexMap = getIndexOnTable(handle, tableName);
```
It looks like you are not taking advantage of the values in this map. Should `getIndexOnTable` just return a list of index names, and then it can be called `getIndexNamesForTable`

It would be really cool if we could infer if the index we are trying to create is already in the table by looking at the column names, ordinal values, etc. but that seems like overkill and just checking the name seems reasonable

pranavbhole replied:
Yes, I am not using the values here excepts in tests. I just created this map that so that we can use it for validating the indexed columns in future or somewhere for other meta tables. We are currently just checking the index name and creating it regardless of the indexed columns.

suneet-s commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -377,8 +380,20 @@ public void createEntryTable(final String tableName)
+        tableName,
+        "datasource_active",
+        StringUtils.format(
+            "CREATE INDEX idx_%1$s_datasource_active_date ON %1$s(datasource, active)",
```
```suggestion
            "CREATE INDEX idx_%1$s_datasource_active ON %1$s(datasource, active)",
```

pranavbhole replied:
fixed

suneet-s commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -377,8 +380,20 @@ public void createEntryTable(final String tableName)
+        "datasource_active",
+        StringUtils.format(
+            "CREATE INDEX idx_%1$s_datasource_active_date ON %1$s(datasource, active)",
+            tableName
```
Have you considered moving these create index commands to the function `alterEntryTable` below? instead of it being part of the function `createEntryTable`

pranavbhole replied:
We are creating index along with the createTable method for all meta tables, thus putting it along with Create table. 
Ideally We should migrate other create index statements to this new construct, so that we attempt to create the indexes if they are missing esp older installations.

imply-cheddar commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -830,4 +845,73 @@ public Void withHandle(Handle handle)
+            public Void withHandle(Handle handle)
+            {
+              if (tableExists(handle, tableName)) {
+                Map<String, List<String>> indexMap = getIndexOnTable(handle, tableName);
```
Your function signatures are going to cause us to read all of the indexes for all of the tables on every single check.  Not a big deal for 2, but better to be more conscious of the number of DB calls that you are making.

If you adjust the signature of `createIndex` to be a bit more developer friendly (as suggested above), then I think that you can build the exact index name that you expect and check for only that.  Alternatively, you should read the set of indexes once and then reuse that map for each of these calls.

imply-cheddar commented on server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java:
```diff
@@ -830,4 +845,73 @@ public Void withHandle(Handle handle)
+   * @param checkString case-insensitive string which substring of index name, it helps to check the existing index on table
+   * @param indexSQL    SQL statement to create index
+   */
+  public void createIndex(final String tableName, final String checkString, final String indexSQL)
```
This function signature is a bit error-prone for the end user.  It's pretty easy (as noted from the earlier comment on this very PR) to accidentally get the "checkString" and the "indexSQL" wrong.  It should be safe to separate out the different parts of the create index command.  You can make this signature

```
createIndex(String tableName, String indexName, List<String> indexCols)
```

And then you should be able to build the correct SQL for that inside of this method as well as know the exact name of the index that you are trying to create (for use in checking if it already exists)

pranavbhole commented:
Amended the existing commit  and addressed the comments.
1. removed the unwanted part from getIndexOnTable and reusing the createdIndexSet for subsequent createIndex calls. 
2. createIndex now builds the SQL statement and checks for exact case-insensitive match for existing indexes.
3. Added more no fail tests.

@cheddar  @suneet-s  Thank you very much for comments, Please review whenever you get chance.

clintropolis reviewed (approved):
tested with my local debugging cluster using postgres and the index was added :+1:
